### PR TITLE
bootstrap: write texts to a .tmp file first for atomicity

### PR DESCRIPTION
**Pull request by nodakai**
If you are using a hard-linked file as your config.toml, this change will affect the way other instances of the file is modified.
The original version would modify all other instances whereas the new version will leave others unchanged, reducing the ref count by one.

### CODE REVIEW

File: src/bootstrap/configure.py [hunk 1]
```diff
@@ -428,11 +429,18 @@ def configure_section(lines, config):
     else:
         configure_section(sections[section_key], section_config)
 
+@contextlib.contextmanager
+def output(filepath):
```
Comment (kennytm) [inline src/bootstrap/configure.py]: `configure` imports `bootstrap`, so just reuse `bootstrap.output` and remove this def.

**Review by kennytm — COMMENTED** [review]
~~r=me with one nit~~.

File: src/bootstrap/configure.py (hunk 1 above)
Comment (nodakai) [reply src/bootstrap/configure.py]: Great point, fixed it accordingly.

**Comment by kennytm** [comment]
Hold on, I have a concern.

> https://docs.python.org/2/library/os.html#os.rename
>
>  On Windows, if *dst* already exists, OSError will be raised even if it is a file; there may be no way to implement an atomic rename when dst names an existing file.

This means it can't really overwrite a file on Windows. 

Calling `os.remove` isn't a complete fix.

> https://docs.python.org/2/library/os.html#os.remove
>
> On Windows, attempting to remove a file that is in use causes an exception to be raised

On Python 3.3+ we could use [`os.replace`](https://docs.python.org/3/library/os.html#os.replace), but these Python scripts need to support both Python 2.x and 3.x.

Could you perform these fallback measures?

1. If `os.replace` exists, use it
2. Try to delete the original file before renaming.
3. Try copy the content to the original file if all failed.

**Comment by nodakai** [comment]
Crap, [this thing](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4017593) totally slipped my mind, I've been away from Win32 these days...

I've pushed a better Win32 support as you suggested, but I'm not sure if the goal of this patch worths the code length anymore.

This is a WIP as I'm still confirming `shutil.copyfile` won't throw in this use case.

**Comment by alexcrichton** [comment]
Could the file be removed, ignoring errors, prior to creation? I think that'd help avoid the platform-specific goop?

**Comment by stokhos** [comment]
Ping from triage, @nodakai  we haven't heard from you for a while, will you have time to look into this PR in near future?

**Comment by nodakai** [comment]
Just pushed smaller fixes; waiting for the CI...

**Comment by alexcrichton** [comment]
Since we aren't really going for speed here, could remove + write always be used? It seems like that would be the most clear with no version or os checks

**Comment by nodakai** [comment]
@alexcrichton Indeed there's little to be gained from sticking to the really-atomic overwrite. Now we have just two cases.

**Comment by alexcrichton** [comment]
The aspect of windows holding a file open can probably be skipped as well, right? This is just a configuration file and doesn't really have a lot of windows programs holding it open I think?

**Comment by nodakai** [comment]
@alexcrichton Your text editor could be one, when you are watching what the generated file looks like?

**Comment by kennytm** [comment]
At least for `notepad.exe` it does keep the file open.

**Comment by alexcrichton** [comment]
Eh I'll leave this up to you @kennytm

**Pull request merged**